langchain/src/vectorstores/elasticsearch.ts

```diff
@@ -0,0 +1,172 @@
+  | { client: Client }
+  | { clientOptions: ClientOptions }
+) & {
+  indexName?: string;
+};
+
+interface VectorDocument {
+  vector: number[];
+  text: string;
+  metadata?: object;
+}
+
+/**
+ * Wrapper around Elasticsearch as a vector database.
+ */
+export class ElasticSearchStore extends VectorStore {
+  indexName: string;
+
+  client: Client;
+
+  constructor(public embeddings: Embeddings, args: ElasticSearchLibArgs) {
+    super(embeddings, args);
+    this.indexName = args.indexName ?? uuid.v4();
+    if (isClientOptions(args)) {
+      this.client = new Client(args.clientOptions);
+    } else {
+      this.client = args.client;
+    }
+  }
+
+  async addVectors(
+    embeddings: number[][],
+    documents: Document[]
+  ): Promise<void> {
+    if (!(await this.client.indices.exists({ index: this.indexName }))) {
+      await this.client.indices.create({
+        index: this.indexName,
+        body: {
+          mappings: {
+            properties: {
+              vector: {
+                type: "dense_vector",
+                dims: embeddings[0].length,
+              },
+              text: { type: "text" },
+              metadata: { type: "object" },
+            },
+          },
+        },
+      });
+    }
+    const operations = documents.flatMap(
+      (document, index): [BulkOperationContainer, VectorDocument] => [
+        {
+          index: {
+            _index: this.indexName,
+            _id: uuid.v4(),
+          },
+        },
+        {
+          vector: embeddings[index],
+          text: document.pageContent,
+          metadata: document.metadata,
+        },
+      ]
+    );
+    const request: BulkRequest<VectorDocument> = {
+      refresh: true,
+      operations,
+    };
+
+    await this.client.bulk(request);
+  }
+
+  async addDocuments(documents: Document[]): Promise<void> {
+    const texts = documents.map(({ pageContent }) => pageContent);
+    await this.addVectors(
+      await this.embeddings.embedDocuments(texts),
+      documents
+    );
+  }
+
+  async similaritySearchVectorWithScore(
+    query: number[],
+    k: number,
+    filter?: object
+  ): Promise<[Document, number][]> {
+    const searchQuery = {
+      index: this.indexName,
+      query: {
+        script_score: {
+          query: {
+            bool: {
+              must: Object.entries(filter || {}).map(([key, value]) => ({
+                match: {
+                  [`metadata.${key}`]: value,
+                },
+              })),
+            },
+          },
+          script: {
+            source: "cosineSimilarity(params.query_vector, 'vector') + 1.0",
+            params: { query_vector: query },
+          },
+        },
+      },
+      size: k,
+    };
+    const response = await this.client.search<VectorDocument>(searchQuery);
+    return response.hits.hits.map((hit) => [
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      new Document<Record<string, any>>({
```

Comment:
The default type for `Document` should cover this, no need to explicitly state it and require TS suppression.